qemu-shutdown

```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+sleep 300
```

Comment:
While testing updates on KVM I noticed that draining is not always possible because machines are "gone" immediately. I figured this shutdown behaviour here and added a sleep to wait for draining. My question is if this can have any weird side effects or may even cause the `system_powerdown` to never happen in weird situations. What can I try to verify to ensure this change does not break shutdown behaviour as we expect it? Anything to look out for?

Reply:
I would like to have this verified by the SREs. <3

Reply:
> draining is not always possible because machines are "gone" immediately

Can you do draining and then shutdown?

Reply:
Not really. The general entrypoint is deleting the pod. Then we reconcile upon the delete event of the pod. This is due to be able to drain on active and passive deletion events. Active is when we say we want to do it, like on scaling. Passive is when somebody else decides it has to be done, like on rescheduling. So the entrypoint is the pod deletion, which triggers the lifecycle hook, which has this `qemu-shutdown` script configured. Why do you ask?

Reply:
We need to test how masters behave with such long 5 min timeout.

Reply:
I was thinking about this. With this we are going to miss the shutdown occasionally because of the time based blocking. We should add a side car container to the pod like the endpoint updater. This one should watch for the `NodeConfig` status which tells it the guest cluster node got drained. Then the qemu shutdown here can happen via shell out and the `NodeConfig` status can be updated again to signal a definite final state to remove the pod and finish the draining/shutdown.

Reply:
See https://github.com/giantswarm/kvm-operator/blob/master/service/controller/v11/resource/deployment/master_deployment.go#L234-L240.

Reply:
I see. Thanks. I'm asking stupid question here, but does even `system_powerdown` has to happen? I understand quemu process is associated with the pod. The quemu machine is drained. Is it really critical to shutdown it gracefully? Maybe it closes network connections or something.

Reply:
The docs say this. 

> The behavior is similar for a PreStop hook. If the hook hangs during execution, the Pod phase stays in a Terminating state and is killed after terminationGracePeriodSeconds of pod ends. If a PostStart or PreStop hook fails, it kills the Container.

So the sleep here might just be sufficient. @r7vme how should masters behave? What can I test/verify to make sure this is fine?

Reply:
I mean when pod finally dies, associated quemu process dies as well. So there is no garbage left. In worst case scenario machine is not stopped gracefully, which can happen also now.

Reply:
I would like to get some insights from the SREs on why we are doing the qemu shutdown here to understand the impact. Based on the docs I would say this is fine though.

Reply:
BTW I don't think there would be 5 min downtime. It's waiting before shutdown.